### PR TITLE
Add jaeger-query HTTP handler diagnostic logging

### PR DESCRIPTION
**Pull request by albertteoh**
Signed-off-by: albertteoh <albert.teoh@logz.io>

## Which problem is this PR solving?
- Jaeger-query was swallowing json marshalling errors, making it difficult to root-cause issues during development (while adding a new `/metrics` API endpoint).

## Short description of the changes
- Handle `json.Marshal` and `ResponseWriter.Write` errors and log them. Log success at debug level as well.

### CODE REVIEW

**Review by yurishkuro — COMMENTED** [review]
The error paths can me tested

**Comment by albertteoh** [comment]
> The error paths can me tested

Yup, sloppiness on my part. I had a go at writing tests for it, please let me know if there's a more idiomatic approach.

File: cmd/query/app/http_handler.go [hunk 1]
```diff
@@ -454,7 +454,15 @@ func (aH *APIHandler) writeJSON(w http.ResponseWriter, r *http.Request, response
...
+	if err != nil {
+		aH.handleError(w, fmt.Errorf("failed marshalling HTTP response to JSON: %w", err), http.StatusInternalServerError)
+		return
+	}
 	w.Header().Set("Content-Type", "application/json")
-	w.Write(resp)
+	if b, err := w.Write(resp); err != nil {
+		aH.handleError(w, fmt.Errorf("failed writing HTTP response: %w", err), http.StatusInternalServerError)
+	} else {
+		aH.logger.Debug("Successfully wrote HTTP response", zap.Int("bytes", b))
```
Comment (yurishkuro) [inline cmd/query/app/http_handler.go:466]: I'm not a fan of this debug logging, is it necessary? It should be observable from the client side already.

File: cmd/query/app/http_handler_test.go [hunk 2]
```diff
@@ -185,16 +186,59 @@ func TestLogOnServerError(t *testing.T) {
...
+func (h *httpResponseErrWriter) Write([]byte) (int, error) {
+	return 0, fmt.Errorf("failed to write")
+}
+func (h *httpResponseErrWriter) WriteHeader(statusCode int) {}
+func (h *httpResponseErrWriter) Header() http.Header {
+	return http.Header{}
+}
+
+// ErrHandlerFunc injects the httpResponseErrWriter into the HTTP handler.
+type ErrHandlerFunc func(http.ResponseWriter, *http.Request)
```
Comment (yurishkuro) [inline cmd/query/app/http_handler_test.go:202]: This type isn't necessary, you can inline in L262 using http.HandlerFunc

**Review by yurishkuro — APPROVED** [review]

**Comment by albertteoh** [comment]
Thanks for the review!

**Pull request merged**